test/react_client/event_helpers_test.dart

```diff
@@ -0,0 +1,999 @@
+      expect(event.cancelable, isFalse);
+      expect(event.defaultPrevented, isTrue);
+      expect(event.isTrusted, isTrue);
+
+      // Other defaults
+      expect(event.type, 'non-default');
+      expect(event.timeStamp, 100);
+
+      // Null defaults
+      expect(event.currentTarget, isA<InputElement>());
+      expect(event.nativeEvent, 'string');
+      expect(event.eventPhase, 0);
+      expect(event.target, isA<InputElement>());
+
+      // Function properties
+      expect(() => event.stopPropagation(), returnsNormally, reason: 'this defaults to an empty function');
+      expect(() => event.preventDefault(), returnsNormally);
+      expect(event.defaultPrevented, isTrue, reason: 'calling preventDefault sets the value to true.');
+    }
+
+    void testSyntheticEventBaseAfterMerge(SyntheticEvent event) {
+      // Boolean defaults
+      expect(event.bubbles, isFalse);
+      expect(event.cancelable, isTrue);
+      expect(event.defaultPrevented, isFalse);
+      expect(event.isTrusted, isFalse);
+
+      // Other defaults
+      expect(event.type, 'updated non-default');
+      expect(event.timeStamp, 200);
+
+      // Null defaults
+      expect(event.currentTarget, isA<DivElement>());
+      expect(event.nativeEvent, 'updated string');
+      expect(event.eventPhase, 2);
+      expect(event.target, isA<DivElement>());
+
+      // Function properties
+      expect(mergeTestCounter, 0, reason: 'test setup sanity check');
+
+      event.stopPropagation();
+      expect(mergeTestCounter, 1);
+
+      event.preventDefault();
+      expect(mergeTestCounter, 2);
+    }
+
+    group('createSyntheticEvent', () {
+      test('returns normally when no parameters are provided', () {
+        testSyntheticEventDefaults(createSyntheticEvent());
+      });
+
+      test('merges values as expected', () {
+        final element = InputElement()
+          ..name = 'test'
+          ..value = 'test value';
+        final baseEvent = createSyntheticEvent(
+          bubbles: true,
+          cancelable: false,
+          currentTarget: element,
+          defaultPrevented: true,
+          preventDefault: null,
+          stopPropagation: null,
+          eventPhase: 0,
+          isTrusted: true,
+          nativeEvent: 'string',
+          target: element,
+          timeStamp: 100,
+          type: 'non-default',
+        );
+        testSyntheticEventBaseForMergeTests(baseEvent);
+
+        final newElement = DivElement();
+        final newEvent = createSyntheticEvent(
+          baseEvent: baseEvent,
+          bubbles: false,
+          cancelable: true,
+          currentTarget: newElement,
+          defaultPrevented: false,
+          preventDefault: () => mergeTestCounter++,
+          stopPropagation: () => mergeTestCounter++,
+          eventPhase: 2,
+          isTrusted: false,
+          nativeEvent: 'updated string',
+          target: newElement,
+          timeStamp: 200,
+          type: 'updated non-default',
```

Comment:
Also related question: if we specifically wanted to set `type` to `null` here, would `newEvent.type` still end up being `'non-default'` because of the `baseEvent`?

Reply:
Actually looking at it again, I think `type` was a bad example since it has a default value when null. Maybe something like `target` though - that could end up being null - though I'm not sure if it's a common case to want to override something in `baseEvent` with null?

Reply:
You're correct that yep, a consumer couldn't pass `null` in to unset a value. I also thought that it wasn't a common use case and could usually be worked around by not passing in a `baseEvent` and instead just pass along only the values that need to be set. In the future if it did become common, I think a change could be just to add a new parameter to the APIs that allows consumers to not get any defaults at all